### PR TITLE
TLS: add `{verify, verify_peer}` to enable verification

### PR DESCRIPTION
**Pull request by jiahuili430**
When the CouchDB custom (couch) distribution is enabled, OTP 24 and 25,
you will get a warning when using `remsh-tls` or `remsh -t`; but in OTP 26,
this is an error. Because the default value of the verify option is `verify_peer`
instead of `verify_none`. We need to provide the appropriate CA certificates to
pass the validation.

```bash
$ ./dev/remsh-tls
=WARNING REPORT==== 4-May-2023::12:43:28.893022 ===
Description: "Server authenticity is not verified since certificate path validation is not enabled"
     Reason: "The option {verify, verify_peer} and one of the options
              'cacertfile' or 'cacerts' are required to enable this."
```

```bash
$ ./dev/remsh-tls
Could not connect to "node1@127.0.0.1"
```

- Add `{verify, verify_peer}` to enable verification
- Add `certs` in `couch_dist` app to generate proper certificates
- Add `-t`/`--enable-tls` mode in ./dev/run to automatically generate
vm.args, certificates and configuration files
- Add `--no-tls <node>` option to specify node to use TCP only
- Add README and documentation
- Remove TLS certificate generation code from `configure`

Greate thanks to Robert Newson for figuring out how to generate the
correct certificates to pass verification!

Ref:
- https://www.erlang.org/doc/apps/ssl/ssl_distribution.html
- https://www.erlang.org/doc/man/ssl_app.html
- https://www.erlang.org/blog/otp-26-highlights/#ssl-safer-defaults
- https://github.com/rnewson/elixir-certs

## Testing recommendations
```bash
./configure --dev --spidermonkey-version 91 && make && ./dev/run --enable-tls
./configure --dev --spidermonkey-version 91 && make && ./dev/run --enable-tls --no-tls node2,node3

./dev/remsh-tls
(node1@127.0.0.1)1> net_kernel:nodes_info().
{ok,[{'node2@127.0.0.1',
         [{owner,<0.456.0>},
          {state,up},
          {address,
              {net_address,{{127,0,0,1},55170},"127.0.0.1",tcp,inet}},
          {type,normal},
          {in,145},
          {out,164}]},
     {'node3@127.0.0.1',
         [{owner,<0.461.0>},
          {state,up},
          {address,
              {net_address,{{127,0,0,1},55172},"127.0.0.1",tcp,inet}},
          {type,normal},
          {in,141},
          {out,169}]},
     {'remsh17312@127.0.0.1',
         [{owner,<0.1418.0>},
          {state,up},
          {address,
              {net_address,{{127,0,0,1},55203},"127.0.0.1",tls,inet}},
          {type,hidden},
          {in,10},
          {out,15}]}]}
```

## Checklist

- [ ] Code is written and works correctly
- [ ] Changes are covered by tests
- [ ] Any new configurable parameters are documented in `rel/overlay/etc/default.ini`
- [x] Documentation changes were made in the `src/docs` folder
- [ ] Documentation changes were backported (separated PR) to affected branches

### CODE REVIEW

File: rel/overlay/etc/vm.args [hunk 1]
```diff
@@ -82,7 +82,8 @@
 ##      [{certfile, "</path/to/erlserver.pem>"},
 ##       {secure_renegotiate, true}]},
 ##     {client,
-##      [{secure_renegotiate, true}]}].
+##      [{secure_renegotiate, true},
+##       {verify, verify_none}]}].
```
Comment (rnewson) [inline rel/overlay/etc/vm.args:86]: I'm very uncomfortable deliberating disabling security if this gets into the release tarball.

File: src/docs/src/cluster/tls_erlang_distribution.rst [hunk 2]
```diff
@@ -36,10 +36,11 @@ the ``certificate`` and its ``private key``.
     .. code-block:: text
 
         [{server,
-          [{certfile, "</path/to/erlserver.pem>"},
+          [{certfile, "<absolute_path/to/erlserver.pem>"},
```
Comment (rnewson) [inline src/docs/src/cluster/tls_erlang_distribution.rst:39]: removing the leading `/` makes it a relative path, even if the name is clearer.

**Review by rnewson — CHANGES_REQUESTED** [review]
I don't think this is right. We don't fix the TLS warning by weakening security. I think we either do nothing or we need to set this up correctly.

File: rel/overlay/etc/vm.args (hunk 1 above)
Comment (jiahuili430) [reply rel/overlay/etc/vm.args:86]: I tried to use `{verify, verify_peer}`, but get errors about bad certificates, or just hang with `*** ERROR: Shell process terminated!`. So add `{verify, verify_none}` to suppress the warning.
```
*** ERROR: Shell process terminated! (^G to start new job) ***
=NOTICE REPORT==== 6-May-2023::13:58:20.360922 ===
TLS client: In state certify at ssl_handshake.erl:2080 generated CLIENT ALERT: Fatal - Bad Certificate
```

File: src/docs/src/cluster/tls_erlang_distribution.rst (hunk 2 above)
Comment (jiahuili430) [reply src/docs/src/cluster/tls_erlang_distribution.rst]: Add leading `/`, thanks for your review.

File: configure [hunk 3]
```diff
@@ -73,17 +73,43 @@ EOF
...
+[{server, [
+    {certfile, "${rootdir}/dev/erlserver.pem"},
+    {secure_renegotiate, true}
+  ]},
+  {client, [
+    {secure_renegotiate, true},
+    {verify, verify_peer},
+    {cacertfile, "${rootdir}/dev/ca-cert.pem"},
+    {fail_if_no_peer_cert, true},
+    {verify_fun, {
```
Comment (iilyak) [inline configure:95]: Can you also take a look into customize_hostname_check. I don't know when it was introduced. But it seems to be a better way.

Comment (jiahuili430) [reply configure]: I tried `{customize_hostname_check, [{match_fun, public_key:pkix_verify_hostname_match_fun(https)}]}` but it didn't work.

Comment (iilyak) [reply configure]: Did you try making the match_fun always passing just to test that function is available in the OTP we are using.

define and export a function in any module

```
always_true() ->
   fun(_) -> true.
```

Then add configuration `{customize_hostname_check, [{match_fun, your_module:always_true()]}`. If it works, look into how https://github.com/erlang/otp/blob/2c864f2f9ef61750006afde6af33453a11d025cd/lib/public_key/src/public_key.erl#L1296 is defined in recent OTP and compare it with the OTP version we use. Understand the difference. If new OTP version has a fix we can make a conditional based on OTP version used. For older versions we would use our own function (inspired by latest OTP). On newer versions of OTP we would use function provided by OTP.

In any case we shouldn't not disable security by keeping `match_fun, your_module:always_true()` or `{verify, verify_none}`. Our goal is to suppress the warning not to disable security.

Comment (jiahuili430) [reply configure]: The problem with this approach is that the TLS distribution starts earlier than other modules, even earlier than the standard io drivers. If we use custom module function in conf file, it will throw error about `undefined function`.

File: rel/overlay/etc/vm.args [hunk 4]
```diff
@@ -108,9 +109,10 @@
 ## By default that would be "clouseau@127.0.0.1".
 ## Don't forget to override the paths to point to your certificate(s) and key(s)!
 ##
-#-proto_dist couch
-#-couch_dist no_tls '"clouseau@127.0.0.1"'
-#-ssl_dist_optfile <path/to/couch_ssl_dist.conf>
+-proto_dist couch
+-couch_dist no_tls '"clouseau@127.0.0.1"'
+-couch_dist no_tls '"node2@127.0.0.1"'
+-ssl_dist_optfile /Users/jiahuili/src/a/dev/couch_ssl_dist.conf
```
Comment (nickva) [inline rel/overlay/etc/vm.args:115]: These are for testing mostly, we don't want to enable this by default?

Comment (jiahuili430) [reply rel/overlay/etc/vm.args:115]: yes, just for testing.

File: configure [hunk 5]
```diff
@@ -69,22 +69,58 @@ Options:
...
+    {certfile, "${rootdir}/dev/erlserver.pem"},
+    {secure_renegotiate, true}
+  ]},
+  {client, [
+    {secure_renegotiate, true},
+    {verify, verify_peer},
+    {cacertfile, "${rootdir}/dev/ca-cert.pem"},
+    {server_name_indication, "$fqdn"}
+  ]}
+].
```
Comment (nickva) [inline configure:121]: Wonder if it makes sense to use `{fail_if_no_peer_cert, true}` as well. Just guessing that users would go for a mTLS (mutual TLS) setup of some sort. Maybe something like this could work?

```
[{server, [
    {certfile, ".../erlserver.pem"},
    {cacertfile, ".../ca-cert.pem"},
    {verify, verify_peer},
    {secure_renegotiate, true},
    {fail_if_no_peer_cert, true}
  ]},
  {client, [
    {cacertfile, ".../ca-cert.pem"},
    {certfile, ".../erlserver.pem"},
    {verify, verify_peer},
    {secure_renegotiate, true},
    {server_name_indication, "localhost"}
  ]}
].
```

**Comment by rnewson** [comment]
I agree with nick. The value of `--generate-tls-dev-cert` is to demonstrate how to correctly set up a couchdb cluster with TLS for the erlang distribution protocol as much as possible. So the nodes should mutually authenticate each other. The nodes will need a certificate each (with their erlang node name as the Common Name attribute), etc. The only difference between what we do here and a real setup should be the node certificates and the CA that signed them.

I've also suggested elsewhere to move the logic from configure to dev/run as we only know the number of nodes at dev/run time. With that done it will be possible to automatically add the correct arguments to `vm.args` for each node, like we currently do to add the individual node names.

File: configure (hunk 5 above)
Comment (jiahuili430) [reply configure:121]: I'll keep trying, thanks for the suggestion.

File: dev/gen_certs.sh [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+  cd ${devdir}/certs
+  echo "Make sure to use different values for the Common Name (FQDN) in CA certificate and certificate."
+
+  echo "Generate CA Certificate:"
+  openssl genrsa 2048 >ca-key.pem
+  openssl req -new -x509 -nodes -days 3650 -key ca-key.pem -out ca-cert.pem -subj "/CN=couch"
+
+  echo "Generate Server Certificate:"
+  openssl req -newkey rsa:2048 -nodes -days 3650 -keyout s-key.pem -out s-req.pem -subj "/CN=$HOST"
+  openssl x509 -req -days 3650 -set_serial 01 -in s-req.pem -out s-cert.pem -CA ca-cert.pem -CAkey ca-key.pem
```
Comment (rnewson) [inline dev/gen_certs.sh:26]: If these are the per-node certificates I think we only need one certificate (the node is a client and a server depending on viewpoint) with CN of erlang node name

File: dev/remsh-tls [hunk 7]
```diff
@@ -20,7 +20,7 @@ if [ -z $NODE ]; then
 fi
 
 if [ -z $HOST ]; then
-    HOST="127.0.0.1"
+    HOST=$(hostname -f)
```
Comment (rnewson) [inline dev/remsh-tls:23]: What is forcing us away from `127.0.0.1` for this?

File: dev/run [hunk 8]
```diff
@@ -264,6 +278,10 @@ def setup_context(opts, args):
         "auto_ports": opts.auto_ports,
         "locald_configs": opts.locald_configs,
         "with_nouveau": opts.with_nouveau,
+        "enable_tls": opts.enable_tls,
+        "tcp_node": opts.tcp_node,
+        "hostname": socket.gethostname() if opts.enable_tls else "127.0.0.1",
+        "bind_address": "0.0.0.0" if opts.enable_tls else "127.0.0.1",
```
Comment (rnewson) [inline dev/run:284]: why is this necessary? Ideally the dev nodes remain bound only to 127.0.0.1. Is something in openssl forcing this?

Comment (rnewson) [reply dev/run:284]: Do you need to override server_name_indication? I thought the CN once set to the erlang node name would be the right value. did that not work? (`"/CN=node1@127.0.0.1"`, I mean.)

Comment (jiahuili430) [reply dev/run:284]: np, `node1@127.0.0.1` does not work. CN needs to be equal to the `hostname`.
e.g.: If the node name is `ssl_test@localhost`, CN needs to equal `localhost`.

I've tried different CNs, such as 127.0.0.1, {127,0,0,1}, 127.0.1.1, localhost, localhost.local, `$hostname -s/-f`, etc. If the node name contains 127.0.0.1, they just return `hostname_check_failed` in OTP 25.

Comment (rnewson) [reply dev/run:284]: it's more complicated than that. reading `inet_tls_dist.erl` makes it clear that they expect the host name part in the SAN field and the node name in a utf8string as a commonName under a directoryName. I have not yet persuaded openssl to generate a certificate that `inet_tls_dist:cert_nodes/1` is happy with. I think the author(s) of `inet_tls_dist` has made a cryptic choice;

```
%% Look in Extensions, in all subjectAltName:s
%% to find node names in this certificate.
%% Host names are picked up as a subjectAltName containing
%% a dNSName, and the first subjectAltName containing
%% a commonName is the node name.
%%
```

The nearest I can get so far is;

```
 [{'Extension',{2,5,29,17},
                                                     false,
                                                     [{dNSName,"127.0.0.1"},
                                                      {directoryName,{rdnSequence,[[{'AttributeTypeAndValue',{2,5,
                                                                                                              4,3},
                                                                                                             <<12,7,99,111,117,99,104,100,98>>}]]}}]},

```

`cert_nodes` finds the hostname but because the `<<12,7,99,111,117,99,104,100,98>>` is not a `utf8String` tuple it doesn't find the node name.

This is the (bizarre) way I have managed to add the dirName to the certificate. all other methods have failed (like the invitingly simple `-addext`).

```
printf "[SAN]\nsubjectAltName=DNS:127.0.0.1,dirName:dir_ext\n[dir_ext]\nCN=couchdb\n" > ext.cnf
openssl x509 -req -days 3650 -set_serial 01 -in node-req.pem -out node-cert.pem -CA ca-cert.pem -CAkey ca-key.pem -extfile ext.cnf -extensions SAN 
```

Thus far I think the openssl cli is mediocre and you can't actually do everything there that openssl is capable of. It seems you _must_ use openssl conf files to do anything interesting.

This has, so far, been frustrating and illuminating.

Comment (rnewson) [reply dev/run:284]: https://github.com/rlipscombe/elixir-certs/blob/main/certs.exs looks helpful...

Comment (jiahuili430) [reply dev/run:284]: Thank you, Bob.

Comment (rnewson) [reply dev/run:284]: Last bit of the mystery is clearing. There is a variant or extended certificate format specific to OTP. If I use that then the dirName is encoded correctly. openssl is obviously unaware of the OTP-specific `OTPCertificate` format, so it cannot be used to generate certificates for inet_tls_dist that would check both node and host name (for sample erlang node name `node@host`).

```
X509.RDNSequence.new("/CN=Bob")
{:rdnSequence,
 [[{:AttributeTypeAndValue, {2, 5, 4, 3}, <<12, 3, 66, 111, 98>>}]]}

X509.RDNSequence.new("/CN=Bob", :otp)
{:rdnSequence, [[{:AttributeTypeAndValue, {2, 5, 4, 3}, {:utf8String, "Bob"}}]]}
```

The rabbit hole is deep and full of terrors.

Comment (rnewson) [reply dev/run:284]: This diff to https://github.com/rlipscombe/elixir-certs allows me to generate a correct certificate.

```
diff --git a/certs.exs b/certs.exs
index 86fb2d2..e52a436 100644
--- a/certs.exs
+++ b/certs.exs
@@ -78,7 +78,8 @@ defmodule Certs do
         ext_key_usage: ext_key_usage([:serverAuth, :clientAuth]),
         subject_key_identifier: true,
         authority_key_identifier: true,
-        subject_alt_name: subject_alt_name([commonName])
+        subject_alt_name: subject_alt_name([commonName]),
+        subject_alt_name: subject_alt_name([{:directoryName, X509.RDNSequence.new(subject, :otp)}])
       ]
     }
   end
```

yields a certificate that ```
5> inet_tls_dist:cert_nodes(OTPCert).
["server@server"]
```

Obviously the values are wrong, the first `server` should be `couchdb` and the second server should be `127.0.0.1` but those are just strings.

Comment (rnewson) [reply dev/run:284]: what we'd actually need;

```
                                        {'Extension',{2,5,29,17},false,[{dNSName,"127.0.0.1"}]},
                                        {'Extension',{2,5,29,17},
                                                     false,
                                                     [{directoryName,{rdnSequence,[[{'AttributeTypeAndValue',{2,
                                                                                                              5,4,3},
                                                                                                             {utf8String,<<"node1">>}}]]}}]}]},
                  {'SignatureAlgorithm',{1,2,840,10045,4,3,2},asn1_NOVALUE},
                  <<48,69,2,32,105,178,12,49,33,238,187,152,243,237,155,
                    128,86,97,241,117,150,223,147,14,210,204,126,69,46,255,
                    22,12,165,28,214,174,2,33,0,133,247,16,211,97,56,77,
                    164,21,139,198,41,184,243,254,91,25,64,39,242,67,179,
                    228,255,131,202,213,27,38,91,54,36>>}
ok
5>
5> inet_tls_dist:cert_nodes(OTPCert).
["node1@127.0.0.1"]
```

Comment (rnewson) [reply dev/run:284]: forked the elixir-certs repo and added options to generate these things: https://github.com/rnewson/elixir-certs/

Comment (rnewson) [reply dev/run:284]: a remsh using inet_tls_dist but without presenting a client certificate is rejected as expected;

```
erl -boot ./start_ssl -name remsh$$@127.0.0.1 -hidden -proto_dist inet_tls  -remsh node1@127.0.0.1
Erlang/OTP 24 [erts-12.3.2.9] [source] [64-bit] [smp:16:16] [ds:16:16:10] [async-threads:1] [jit]

=WARNING REPORT==== 25-May-2023::13:09:35.881585 ===
Description: "Authenticity is not established by certificate path validation"
     Reason: "Option {verify, verify_peer} and cacertfile/cacerts is missing"

*** ERROR: Shell process terminated! (^G to start new job) ***
=NOTICE REPORT==== 25-May-2023::13:09:36.033411 ===
TLS client: In state cipher received SERVER ALERT: Fatal - Handshake Failure

(node1@127.0.0.1)1> =NOTICE REPORT==== 25-May-2023::13:09:36.023069 ===
TLS server: In state certify at tls_dtls_connection.erl:315 generated SERVER ALERT: Fatal - Handshake Failure
 - no_client_certificate_provided
```

Comment (jiahuili430) [reply dev/run:284]: Thanks for all the info, especially for generating the correct certificates to pass validation!

File: dev/run [hunk 9]
```diff
@@ -524,6 +544,20 @@ def hack_local_ini(ctx, contents):
     return contents + "\n\n[chttpd_auth]\nsecret = %s\n" % COMMON_SALT
 
 
+def hack_vm_args(ctx, node, contents):
+    contents += f"""
+-proto_dist couch
+-couch_dist no_tls '"clouseau@127.0.0.1"'
```
Comment (rnewson) [inline dev/run:550]: each node will talk to a different clouseau (clouseau1 for node1, clouseau2 for node2, etc)

Comment (jiahuili430) [reply dev/run]: Changed, thanks for the review!

File: dev/remsh-tls [hunk 10]
```diff
@@ -24,6 +24,6 @@ if [ -z $HOST ]; then
 fi
 
 NAME="remsh$$@$HOST"
```
Comment (rnewson) [inline dev/remsh-tls:26]: will we need to generate a remsh certificate dynamically now?

Comment (jiahuili430) [reply dev/remsh-tls:26]: `erl -name $NAME -remsh "$NODE@$HOST" -hidden -proto_dist inet_tls -ssl_dist_optfile "src/couch_dist/certs/out/${NODE}-dist.conf"`

Maybe not, since we're using a dynamic `${NODE}-dist.conf` and certificates.

Comment (rnewson) [reply dev/remsh-tls:26]: despite the code in `inet_tls_dist.erl` that looks for the dirname entry (and parses it out in cert_nodes/1), it appears to be ignored.

It suffices to have a subjectAltName of the hostname. My elixir-certs fork is unnecessary and should not be used.

The cert resulting from

```
./certs create-cert --issuer-cert out/ca-cert.pem --issuer-key out/ca-key.pem --out-cert out/node-cert.pem --out-key out/node-key.pem --template server --subject '/CN=127.0.0.1'```

is sufficient and can be used as the client and server cert of all nodes with the same hostname, which is all of the dev nodes.

Obviously a 'real' cluster would have distinct ip's or hostnames for each node, though.

Comment (rnewson) [reply dev/remsh-tls:26]: So

```
commit 5586959ed008c09141689f1e8865476150e48519
Author: Raimo Niskanen <raimo@erlang.org>
Date:   Thu Apr 26 09:37:24 2018 +0200

    Allow check for node name
```

I think introduced the ability to check the node names also, from the data in the certificate

and

```
commit 794df8cbba8d7942dcb3bf2cbdfa526b04d41dd3
Author: Raimo Niskanen <raimo@erlang.org>
Date:   Fri Jun 8 15:49:06 2018 +0200

    Use public_key to verify client hostname
```
removed it (and used the net_kernel:allow list to restrict node names instead).

they just left the cert_nodes function in but it's unreachable?

Comment (jiahuili430) [reply dev/remsh-tls:26]: May I keep your `elixir-certs` fork in this demo, as it might be used for a `real` cluster setup?
I can modify the `gen_cert` script to use the `server` template and point to one conf and certificate.

Comment (rnewson) [reply dev/remsh-tls:26]: I think it was maybe optional previously with ```{verify_fun, {inet_tls_dist, verify_client, undefined}},``` option, but I am getting confused at this point.

Jessica and I have independently shown that 'node1@127.0.0.1' can start up with node2-dist.conf (which points at node2-cert.pem which has the dirname of 'node2') and still ping 'node2@127.0.0.1' which uses node2-dist.conf also.

So erlang does not check the node name embedded in the cert and apparently cannot be configured to do so.

the `inet_tls_dist:cert_nodes/1` function (and parse_extensions) just exists to tease me.

Comment (rnewson) [reply dev/remsh-tls:26]: @RaimoNiskanen in case you might find our journey amusing ^.

File: dev/run [hunk 11]
```diff
@@ -234,6 +234,20 @@ def get_args_parser():
...
+    parser.add_option(
+        "-t",
+        "--enable-tls",
+        dest="enable_tls",
+        default=False,
+        action="store_true",
+        help="Enable custom TLS distribution (couch)",
+    )
+    parser.add_option(
+        "--tcp-node",
```
Comment (rnewson) [inline dev/run:246]: since this is setting the 'no_tls' configuration parameter, which can match multiple nodes, I think `no_tls` is a better description than `tcp-node`. Also the TLS connections will use TCP too, so this name doesn't distinguish the two kinds of node connection.

File: dev/remsh-tls [hunk 12]
```diff
@@ -26,4 +26,4 @@ fi
...
 NODE="node$NODE@$HOST"
 rootdir="$(cd "${0%/*}" 2>/dev/null; echo "$PWD")"
-erl -name $NAME -remsh $NODE -hidden -proto_dist inet_tls -ssl_dist_optfile "${rootdir}/couch_ssl_dist.conf"
+erl -name $NAME -remsh $NODE -hidden -proto_dist inet_tls -ssl_dist_optfile "${rootdir}/../src/couch_dist/certs/out/couch_dist.conf"
```
Comment (rnewson) [inline dev/remsh-tls:29]: consider passing the location in (env var, perhaps) rather than having to do `../src`.

File: src/couch_dist/README.md [hunk 13]
```diff
@@ -0,0 +1,160 @@
+# couch_dist
+
+`couch_dist` implements a custom distribution protocol -- `couch`, which allows
+nodes to connect using different protocols, such as Transport Layer Security
+(TLS) and Transmission Control Protocol (TCP).
```
Comment (rnewson) [inline src/couch_dist/README.md:5]: Erlang communicates with its own protocol over TCP. It can also be configured to run its protocol over a TLS connection which itself is over TCP.

File: src/couch_dist/README.md [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+      -proto_dist couch
+      -couch_dist no_tls '"clouseau@127.0.0.1"'
+      -ssl_dist_optfile </absolute/path/to/couch_dist.conf>
+      ```
+
+2. `couch_dist.conf`:
+
+    - `erlserver.pem`: contains the certificate and its private key.
+    - `{verify, verify_peer}`: you can specify the hostname with `{server_name_indication, <hostname>}`.
+    - `{fail_if_no_peer_cert, true}`: should be used on the server side only,
```
Comment (rnewson) [inline src/couch_dist/README.md:31]: `fail_if_no_peer_cert` should be set on client and server.

File: src/couch_dist/README.md [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+      ```vm.args
+      -proto_dist couch
+      -couch_dist no_tls '"clouseau@127.0.0.1"'
+      -ssl_dist_optfile </absolute/path/to/couch_dist.conf>
+      ```
+
+2. `couch_dist.conf`:
+
+    - `erlserver.pem`: contains the certificate and its private key.
+    - `{verify, verify_peer}`: you can specify the hostname with `{server_name_indication, <hostname>}`.
```
Comment (rnewson) [inline src/couch_dist/README.md:30]: we are showing how to generate correct certificates, so server_name_indication should not be necessary.

File: src/couch_dist/README.md [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+  --issuer-cert ca-cert.pem --issuer-key ca-key.pem \
+  --out-cert cert.pem --out-key key.pem \
+  --template server \
+  --subject "/CN=127.0.0.1"
+
+# Generate `erlserver.pem`
+cat key.pem cert.pem >erlserver.pem
+
+# Parse certificate to verify:
+# {server_name_indication, "<hostname>"}
```
Comment (rnewson) [inline src/couch_dist/README.md:79]: server_name_indication is not needed.

File: src/couch_dist/gen_certs [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+    {verify, verify_peer},
+    {fail_if_no_peer_cert, true}
+  ]},
+  {client, [
+    {cacertfile, "$(pwd)/out/ca-cert.pem"},
+    {certfile,   "$(pwd)/out/cert.pem"},
+    {keyfile,    "$(pwd)/out/key.pem"},
+    {secure_renegotiate, true},
+    {verify, verify_peer},
+    {server_name_indication, "127.0.0.1"}
```
Comment (rnewson) [inline src/couch_dist/gen_certs:41]: should not be necessary

File: src/docs/src/cluster/tls_erlang_distribution.rst [hunk 18]
```diff
@@ -29,28 +29,55 @@ Reference: `Using TLS for Erlang Distribution`_
...
+        $ ./certs self-signed \
+            --out-cert ca-cert.pem --out-key ca-key.pem \
+            --template root-ca \
+            --subject '/CN=CouchDB Root CA'
+        $./certs create-cert \
+            --issuer-cert ca-cert.pem --issuer-key ca-key.pem \
+            --out-cert cert.pem --out-key key.pem \
+            --template node \
+            --subject '/OU=CouchDB' \
+            --host hostname \
```
Comment (rnewson) [inline src/docs/src/cluster/tls_erlang_distribution.rst:71]: earlier we used the original 'server' template but here we use my 'node' template, which turns out not to be necessary. let's be consistent in our examples and document the use of the 'server' template here also.

File: dev/run (hunk 11 above)
Comment (jiahuili430) [reply dev/run]: Changed to `--no-tls`

File: src/couch_dist/README.md (hunk 14 above)
Comment (jiahuili430) [reply src/couch_dist/README.md:34]: The connection succeeds on OTP 24 and 25 but fails on OTP 26.
`./dev/remsh-tls` fails to connect to node1 if `{fail_if_no_peer_cert, true}` is present on the client side.

See https://www.erlang.org/blog/otp-26-highlights/#ssl-improved-checking-of-options

File: src/couch_dist/README.md (hunk 15 above)
Comment (jiahuili430) [reply src/couch_dist/README.md]: Removed

File: src/couch_dist/README.md (hunk 16 above)
Comment (jiahuili430) [reply src/couch_dist/README.md]: removed

File: src/docs/src/cluster/tls_erlang_distribution.rst (hunk 18 above)
Comment (jiahuili430) [reply src/docs/src/cluster/tls_erlang_distribution.rst]: Change back to `server` template, thanks for reviewing!

File: src/couch_dist/gen_certs (hunk 17 above)
Comment (jiahuili430) [reply src/couch_dist/gen_certs]: removed

File: src/couch_dist/README.md (hunk 14 above)
Comment (jiahuili430) [reply src/couch_dist/README.md]: How about adding this to the `gen_cert` script?
Generate different `couch_dist.conf` files for different erlang versions.
```bash
ERL_VER=$(erl -eval "io:put_chars(erlang:system_info(otp_release)), halt()." -noshell)
if [ $ERL_VER -ge "26" ]; then
  OPTS="{verify, verify_peer}"
else
  OPTS="{verify, verify_peer},
  {fail_if_no_peer_cert, true}"
fi
```

Comment (jiahuili430) [reply src/couch_dist/README.md]: > In OTP 26, the checking of options is strengthened to return errors for incorrect options that used to be silently ignored. For example, ssl now rejects the fail_if_no_peer_cert option if used for the client:

> In OTP 25, the option would be silently ignored.

https://www.erlang.org/blog/otp-26-highlights/#ssl-improved-checking-of-options

Or just leave `{fail_if_no_peer_cert, true}` on the server side. Because when we specify `{verify, verify_peer}`, it requires us to provide cacertfile, certfile, and keyfile to connect successfully.

Comment (jiahuili430) [reply src/couch_dist/README.md]: Revert `gen_cert` script back, so `{fail_if_no_peer_cert, true}` would only appear on server config side.
Also removed `couch_dist` from `couch.app.src` file, and tested it with OTP 24/25/26, worked on my end.

**Review by rnewson — APPROVED** [review]
looks good to me, great job!
